modules/network/src/main/java/org/apache/ignite/network/scalecube/ScaleCubeDirectMarshallerTransport.java

```diff
@@ -0,0 +1,243 @@
+
+            LOG.info("[{0}][doStop] Stopped", address);
+            return Mono.empty();
+        });
+    }
+
+    /** {@inheritDoc} */
+    @Override public Address address() {
+        return address;
+    }
+
+    /** {@inheritDoc} */
+    @Override public Mono<Transport> start() {
+        return Mono.just(this);
+    }
+
+    /** {@inheritDoc} */
+    @Override public Mono<Void> stop() {
+        return doStop();
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean isStopped() {
+        return onStop.isDisposed();
+    }
+
+    /** {@inheritDoc} */
+    @Override public Mono<Void> send(Address address, Message message) {
+        var addr = InetSocketAddress.createUnresolved(address.host(), address.port());
+        return Mono.defer(() -> Mono.fromFuture(connectionManager.channel(addr)))
+            .flatMap(client -> {
+                client.send(fromMessage(message));
+                return Mono.empty();
+            });
+    }
+
+    /**
+     * Handles new network messages from {@link #connectionManager}.
+     *
+     * @param source Message source.
+     * @param msg Network message.
+     */
+    private void onMessage(InetSocketAddress source, NetworkMessage msg) {
+        Message message = fromNetworkMessage(msg);
+
+        if (message != null)
+            sink.next(message);
+    }
+
+    /**
+     * Wrap ScaleCube {@link Message} with {@link NetworkMessage}.
+     *
+     * @param message ScaleCube message.
+     * @return Netowork message that wraps ScaleCube message.
+     * @throws IgniteInternalException If failed to write message to ObjectOutputStream.
+     */
+    private NetworkMessage fromMessage(Message message) throws IgniteInternalException {
+        Object dataObj = message.data();
+        ByteArrayOutputStream stream = new ByteArrayOutputStream();
+
+        try(ObjectOutputStream oos = new ObjectOutputStream(stream)) {
+            oos.writeObject(dataObj);
+        }
+        catch (IOException e) {
+            throw new IgniteInternalException(e);
+        }
+
+        return new ScaleCubeMessage(stream.toByteArray(), message.headers());
+    }
+
+    /**
+     * Unwrap ScaleCube {@link Message} from {@link NetworkMessage}.
+     *
+     * @param networkMessage Network message.
+     * @return ScaleCube message.
+     * @throws IgniteInternalException If failed to read ScaleCube message byte array.
+     */
+    private Message fromNetworkMessage(NetworkMessage networkMessage) throws IgniteInternalException {
+        if (networkMessage instanceof ScaleCubeMessage) {
+            ScaleCubeMessage msg = (ScaleCubeMessage) networkMessage;
+
+            Map<String, String> headers = msg.getHeaders();
+
+            Object obj;
+
+            try(ObjectInputStream ois = new ObjectInputStream(new ByteArrayInputStream(msg.getArray()))) {
+                obj = ois.readObject();
+            }
+            catch (Exception e) {
+                throw new IgniteInternalException(e);
```

Comment:
How will scalecube react to IgniteInternalException? I think we need something else.

Reply:
Weeeell, it shouldn't happen in the first place, because all ScaleCube messages are serializable. I don't really know what we can do here. It's a failfast thingy, should fail our tests